src/operator/nn/cudnn/cudnn_batch_norm-inl.h

```diff
@@ -262,15 +262,26 @@ class CuDNNBatchNormOp {
 
  private:
   void Init(const TBlob &in_data) {
-    if (in_data.ndim() == 4) {
-      for (int i = 0; i < 4; ++i)
-        shape_[i] = in_data.shape_[i];
+    CHECK_GE(param_.axis, 0);
+    CHECK_LT(param_.axis, in_data.ndim());
+    if (param_.axis == 1) {
+      if (in_data.ndim() == 4) {
+        for (int i = 0; i < 4; ++i)
+          shape_[i] = in_data.shape_[i];
+      } else {
+        // when in_data.ndim() != 4
+        shape_[0] = in_data.shape_[0];
+        shape_[1] = in_data.ndim() > 1 ? in_data.shape_[1] : 1;
+        shape_[2] = 1;
+        shape_[3] = in_data.shape_.ProdShape(2, in_data.ndim());
+      }
     } else {
-      // when in_data.ndim() != 4
-      shape_[0] = in_data.shape_[0];
-      shape_[1] = in_data.ndim() > 1 ? in_data.shape_[1] : 1;
+      // reshape to (N, C, 1, D), C is the `param_.axis` dimension
+      shape_[0] = static_cast<dim_t>(in_data.shape_.ProdShape(0, param_.axis));
+      shape_[1] = in_data.shape_[param_.axis];
       shape_[2] = 1;
-      shape_[3] = in_data.shape_.ProdShape(2, in_data.ndim());
+      shape_[3] = static_cast<dim_t>(in_data.shape_.ProdShape(param_.axis + 1,
+            static_cast<int>(in_data.ndim())));
```

Comment:
why do we need static_cast here? why cant we do it like line 276?

Reply:
The return dtype of `ProdShape` is `size_t`, a unsigned int, but the dtype of shape_[i] is `dim_t`, namely `int64_t`.
I do not know whether the `static_cast` is neceseary to avoid the potential compiler's warning.

Reply:
Can we add static_cast to line 276 as well?

Reply:
in_data.ndims() shouldn’t need a static cast though, right?

Reply:
Yes, the dtype of `.ndim()` is `int32_t`.

variable|dtype
---|---
ProdShape | size_t
shape_[i] | dim_t, int64_t
ndim() | int32_t

Reply:
Hi @mseth10 , do you have any suggestion about whether to use `static_cast` ?

Reply:
Imo you can ignore the compiler warnings (if any) for int32_t to int,
you can keep static_cast for size_t to dim_t

Reply:
I see. I have updated it : )